File: packages/core/src/requirement-files.ts
/**
 * GRD-SYS-007: allowed requirement file suffixes (longest first so `.req.yaml` matches before `.req.yml`).
 */
export const REQUIREMENT_FILE_EXTENSIONS = [".req.yaml", ".req.yml"] as const;

/** Default suffix for new files (bootstrap, VS Code new requirement). */
export const REQUIREMENT_FILE_EXTENSION = ".req.yml";

export function isRequirementFilename(basename: string): boolean {
  return REQUIREMENT_FILE_EXTENSIONS.some((ext) => basename.endsWith(ext));
}

/** Strip the requirement extension; returns null if the basename is not a requirement file. */
export function requirementIdFromFilename(basename: string): string | null {
  for (const ext of REQUIREMENT_FILE_EXTENSIONS) {
    if (basename.endsWith(ext)) {
      return basename.slice(0, -ext.length);
    }
  }
  return null;
}

/** For validation error messages. */
export function requirementFileExtensionsDisplay(): string {
  return REQUIREMENT_FILE_EXTENSIONS.join(" or ");
}

/** Valid basenames for a given id, for error messages. */
export function expectedRequirementBasenamesForId(id: string): string {
  return REQUIREMENT_FILE_EXTENSIONS.map((ext) => `${id}${ext}`).join(" or ");
}

File: crates/shallgraph-core/src/requirement_files.rs
//! GRD-SYS-007: allowed requirement file suffixes (longest first so `.req.yaml` matches before `.req.yml`).

pub const REQUIREMENT_FILE_EXTENSIONS: [&str; 2] = [".req.yaml", ".req.yml"];

/// Default suffix for new files (bootstrap).
pub const REQUIREMENT_FILE_EXTENSION: &str = ".req.yml";

pub fn is_requirement_filename(basename: &str) -> bool {
    REQUIREMENT_FILE_EXTENSIONS
        .iter()
        .any(|ext| basename.ends_with(ext))
}

/// Strip the requirement extension; returns None if the basename is not a requirement file.
pub fn requirement_id_from_filename(basename: &str) -> Option<&str> {
    for ext in REQUIREMENT_FILE_EXTENSIONS {
        if let Some(stem) = basename.strip_suffix(ext) {
            return Some(stem);
        }
    }
    None
}

/// For validation error messages.
pub fn requirement_file_extensions_display() -> String {
    REQUIREMENT_FILE_EXTENSIONS.join(" or ")
}

/// Valid basenames for a given id, for error messages.
pub fn expected_requirement_basenames_for_id(id: &str) -> String {
    REQUIREMENT_FILE_EXTENSIONS
        .iter()
        .map(|ext| format!("{id}{ext}"))
        .collect::<Vec<_>>()
        .join(" or ")
}
